das/daser.go

```diff
@@ -90,5 +149,91 @@ func (d *DASer) sampling(ctx context.Context, sub header.Subscription) {
 		sampleTime := time.Since(startTime)
 		log.Infow("sampling successful", "height", h.Height, "hash", h.Hash(),
 			"square width", len(h.DAH.RowsRoots), "finished (s)", sampleTime.Seconds())
+
+		checkpoint = h.Height
+	}
+}
+
+// catchUpJob represents a catch-up job. (from:to]
+type catchUpJob struct {
+	from, to int64
+}
+
+// catchUpScheduler spawns and manages catch-up jobs, exiting only once all jobs
+// are complete and the last known DASing checkpoint has been stored to disk.
+func (d *DASer) catchUpScheduler(ctx context.Context, checkpoint int64) {
+	wg := sync.WaitGroup{}
+
+	defer func() {
+		// wait for all catch-up jobs to finish
+		wg.Wait()
+		// store latest DASed checkpoint to disk here to ensure that if DASer is not yet
+		// fully caught up to network head, it will resume DASing from this checkpoint
+		// up to current network head
+		// TODO @renaynay: Implement Share Cache #180 to ensure no duplicate DASing over same
+		//  header
+		if err := storeCheckpoint(d.cstore, checkpoint); err != nil {
+			log.Errorw("storing checkpoint to disk", "height", checkpoint, "err", err)
+		}
+		// signal that all catchUp routines have finished
+		close(d.catchUpDn)
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case job := <-d.jobsCh:
+			// spawn catchUp routine
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				d.catchUp(ctx, job)
+				// TODO @renaynay: assumption that each subsequent job is to a higher height than the previous.
+				//  I don't see why that is *not* the case, though.
+				checkpoint = job.to
```

Comment:
The assumption is correct. The problem is that we cannot guarantee the order in which catchups are finished. Probably the last bug I see in the code. If there are multiple catchUp routines spawned on loc 189. All of them will conflict on writing on this line. This is also racy and if we tested a case with multiple catchUp routines with race detector on, then it will complain.

Reply:
We could probably solve this by introducing catchUp job IDs. Do you think this needs to be solved in this PR or can we solve it in a PR that implements job IDs + solves this race condition?